### PR TITLE
authz: support ON_WITH_INCLUSION and ON_WITH_EXCLUSION mode

### PR DESCRIPTION
**Pull request by yangminzhu**
See #6377

### CODE REVIEW

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 1]
```diff
@@ -81,11 +77,18 @@ func (Plugin) OnInboundListener(in *plugin.InputParams, mutable *plugin.MutableO
 	if in.Node.Type != model.Sidecar || in.ListenerType != plugin.ListenerTypeHTTP {
 		return nil
 	}
-	if !isRbacEnabled(in.Env.IstioConfigStore) {
+	hostname := in.ServiceInstance.Service.Hostname
+	attr, err := in.Env.GetServiceAttributes(hostname)
```
Comment (quanjielin) [inline pilot/pkg/networking/plugin/authz/rbac.go]: check err ?

File: pilot/pkg/model/authorization.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+// limitations under the License.
+
+package model
+
+const (
+	// RbacConfigName is the name of the RbacConfig custom resource that controls the RBAC behavior.
+	RbacConfigName = "rbac-config"
+
+	// RbacConfigNamespace is the namespace of the RbacConfig custom resource that controls the RBAC behavior.
+	RbacConfigNamespace = "istio-system"
```
Comment (diemtvu) [inline pilot/pkg/model/authorization.go]: Customer can chose to install Istio on different namespace. I'm not sure using hard-coded value is a good idea.
Also, do you consider to use cluster CRD instead, for global/mesh config. We did it for authn recently, and framework is more ready for you to do so now.

File: pilot/pkg/model/validation.go [hunk 3]
```diff
@@ -1807,14 +1807,14 @@ func ValidateServiceRoleBinding(name, namespace string, msg proto.Message) error
...
-	if !ok {
+	if _, ok := msg.(*rbac.RbacConfig); !ok {
 		return errors.New("cannot cast to RbacConfig")
 	}
 
-	switch in.Mode {
-	case rbac.RbacConfig_ON_WITH_EXCLUSION, rbac.RbacConfig_ON_WITH_INCLUSION:
-		return errors.New("rbac mode not implemented, currently only supports ON/OFF")
+	// TODO(yangminzhu): Don't check the namespace in validation to avoid breaking e2e test, consider
+	// make the RbacConfig to be cluster scope.
```
Comment (diemtvu) [inline pilot/pkg/model/validation.go]: I would consider this move now :)

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 4]
```diff
@@ -113,46 +116,57 @@ func (Plugin) OnOutboundCluster(env model.Environment, node model.Proxy, service
...
+		if namespace == ns {
+			return true
+		}
+	}
+	for _, service := range li.Services {
+		if service == svc {
+			return true
+		}
+	}
+	return false
```
Comment (rshriram) [inline pilot/pkg/networking/plugin/authz/rbac.go]: Is this right? Don’t you need to match both service and namespace before returning true ?

Comment (yangminzhu) [reply pilot/pkg/networking/plugin/authz/rbac.go]: This is intended. The namespace has a larger scope, the semantic is to include all services in that `namespace` field.

File: pilot/pkg/model/validation.go (hunk 3 above)
Comment (yangminzhu) [reply pilot/pkg/model/validation.go]: Done.

File: pilot/pkg/model/authorization.go (hunk 2 above)
Comment (yangminzhu) [reply pilot/pkg/model/authorization.go]: Updated to cluster scope and also removed this hard code.

File: pilot/pkg/model/authorization.go [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+const (
+	// DefaultRbacConfigName is the name of the cluster-scoped rbac config. Only RbacConfig with this
```
Comment (liminw) [inline pilot/pkg/model/authorization.go]: cluster-scoped -> mesh global?

File: pilot/pkg/networking/plugin/authz/rbac.go [hunk 6]
```diff
@@ -113,46 +116,57 @@ func (Plugin) OnOutboundCluster(env model.Environment, node model.Proxy, service
 	servicePort *model.Port, cluster *xdsapi.Cluster) {
 }
 
-func isRbacEnabled(store model.IstioConfigStore) bool {
+// isServiceInList checks if a given service of namespace is found in the RbacConfig target list.
```
Comment (liminw) [inline pilot/pkg/networking/plugin/authz/rbac.go]: a given service "of" namespace -> a given service "or" namespace.

File: pilot/pkg/networking/plugin/authz/rbac.go (hunk 4 above)
Comment (rshriram) [reply pilot/pkg/networking/plugin/authz/rbac.go]: Yes i understand that. But the second if implies that when namespace does not match, match by service name atleast. That seems odd. 
You typically match foo.ns.svc.cluster.local or *.ns.svc...

The above logic does *.ns.svc and then foo.*.svc

Comment (yangminzhu) [reply pilot/pkg/networking/plugin/authz/rbac.go]: I see, here we're actually using full qualified service name.
So a typical target like this:
```
RbacConfig_Target {
Service: ["product.product-ns.svc.cluster.local", "database.database-ns.svc.cluster.local"]
Namespace: ["default"]
}
```
enables all services in "default" plus another 2 services (`product` and `database`)

File: pilot/pkg/model/authorization.go [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package model
+
+const (
+	// DefaultRbacConfigName is the name of the mesh global RbacConfig name. Only RbacConfig with this
+	// name will be considered.
+	DefaultRbacConfigName = "rbac-config"
```
Comment (diemtvu) [inline pilot/pkg/model/authorization.go]: May be "default" instead?

File: pilot/pkg/model/config.go [hunk 8]
```diff
@@ -1113,12 +1114,16 @@ func (store *istioConfigStore) ServiceRoleBindings(namespace string) []Config {
...
 
-func (store *istioConfigStore) RbacConfig(name, namespace string) *Config {
-	rbacConfigs, err := store.List(RbacConfig.Type, namespace)
+func (store *istioConfigStore) RbacConfig(name string) *Config {
+	rbacConfigs, err := store.List(RbacConfig.Type, "")
 	if err != nil {
 		log.Errorf("failed to get rbacConfig: %v", err)
 		return nil
 	}
+	if len(rbacConfigs) > 1 {
```
Comment (diemtvu) [inline pilot/pkg/model/config.go]: Would it be better to ignore all, except the one with name == DefaultRbacConfigName ? You should not have one with name other than DefaultRbacConfigName  anyway, as validation should catch it. Anyway, ignore those should be safe enough.

File: pilot/pkg/model/config_test.go [hunk 9]
```diff
@@ -891,26 +891,20 @@ func TestServiceRoleBindings(t *testing.T) {
 
 func TestRbacConfig(t *testing.T) {
 	store := model.MakeIstioStore(memory.Make(model.IstioConfigTypes))
-	addRbacConfigToStore(model.RbacConfig.Type, "rbac-config", "istio-system", store, t)
+	addRbacConfigToStore(model.RbacConfig.Type, "rbac-config", "", store, t)
```
Comment (diemtvu) [inline pilot/pkg/model/config_test.go]: use constant (DefaultRbacConfigName) instead of "rbac-config"

File: pilot/pkg/model/config.go [hunk 10]
```diff
@@ -1113,12 +1114,16 @@ func (store *istioConfigStore) ServiceRoleBindings(namespace string) []Config {
 	return bindings
 }
 
-func (store *istioConfigStore) RbacConfig(name, namespace string) *Config {
-	rbacConfigs, err := store.List(RbacConfig.Type, namespace)
+func (store *istioConfigStore) RbacConfig(name string) *Config {
```
Comment (diemtvu) [inline pilot/pkg/model/config.go]: Is `name` necessary?

File: pilot/pkg/model/config.go [hunk 11]
```diff
@@ -526,14 +526,15 @@ var (
...
-	// RbacConfig describes an global RBAC config.
+	// RbacConfig describes the mesh level RBAC config.
 	RbacConfig = ProtoSchema{
-		Type:        "rbac-config",
-		Plural:      "rbac-configs",
-		Group:       "config",
-		Version:     istioAPIVersion,
-		MessageName: "istio.rbac.v1alpha1.RbacConfig",
-		Validate:    ValidateRbacConfig,
+		ClusterScoped: true,
```
Comment (diemtvu) [inline pilot/pkg/model/config.go]: Have you try 'upgrade' Istio from previous version to this? As you already define this CRD in previous release, you might not be able to change the scoped, without remove the old one. I'm not sure what the best way for migration, May be add smt to the installation template to remove old CRD (assuming no one using this yet) before (re)install Istio.

Comment (yangminzhu) [reply pilot/pkg/model/config.go]: The way I 'upgrade' Istio is simply remove the existing one and then install the newer version. And I think there is no one really using this CRD, even in test. I didn't change other fields but only the ClusterScoped, is this considered unsafe?

File: pilot/pkg/model/config.go (hunk 8 above)
Comment (yangminzhu) [reply pilot/pkg/model/config.go]: Done, only do a log for this.

File: pilot/pkg/model/config_test.go (hunk 9 above)
Comment (yangminzhu) [reply pilot/pkg/model/config_test.go]: Done

File: pilot/pkg/model/config.go (hunk 10 above)
Comment (yangminzhu) [reply pilot/pkg/model/config.go]: Removed

File: pilot/pkg/model/authorization.go (hunk 7 above)
Comment (yangminzhu) [reply pilot/pkg/model/authorization.go]: Done

File: pilot/pkg/model/config.go (hunk 11 above)
Comment (yangminzhu) [reply pilot/pkg/model/config.go]: Thanks for the remind, Diem. I'll test it again and probably add some notes to the doc, better to remove the old RbacConfig when upgrading, but need to investigate how to do this.

**Pull request merged**